_posts/2020-09-21-porting-namecoin-in-tor-browser-to-macos.md

```diff
@@ -0,0 +1,46 @@
+
+Alas, the resulting binaries consistently failed to detect Electrum-NMC and ncprop279.
+
+## What is this external user data craziness?
+
+It turns out that Tor Browser for macOS does something very different from Tor Browser for GNU/Linux and Windows.  On macOS, the user data directory is stored separately from the application executable code.  To make matters worse, on macOS there is no way to supply "default" contents for the user data directory.  In fact, in order to set the default bookmarks for Firefox, Tor Browser on macOS does some kind of witchcraft involving stuffing the bookmarks file inside one of the Firefox application archives, such that when Firefox eventually creates a fresh profile, the bookmarks end up in the desired place.  For non-Firefox data, there simply is no mechanism at all.  Even the file `torrc`, which exists in the source code, never actually gets installed on macOS (which I suppose is tolerated because it's an empty file, so who cares if it gets installed?).
+
+Alas, the Namecoin support in Tor Browser relies on supplying default config files for both Electrum-NMC and ncprop279.  Fixing this was going to take some refactoring.
+
+## Electrum-NMC Support
+
+I considered a few options here.  Initially the thought occurred to me of patching Electrum-NMC to support a read-only config file in a separate directory from its writeable data directory.  Alas, this seemed rather invasive, and I didn't want to write a complex patch without good reason.  I also considered generating the config file on-demand as part of Tor Launcher, but this was also not going to be very clean.  Instead, I realized that we could just pass the config options we cared about as command-line flags, not supply a default config file at all, and be done with it.
+
+Oh wait.  Murphy says no.  Turns out that the RPC port settings in Electrum can't be set on the command-line, only in the config file.  Upon consulting with SomberNight from upstream Electrum, he pointed me to a spot in the relevant Python source, where I could probably do a 3-line patch that would do what I want, and would probably be upstreamable without controversy.  This ended up working very well.  We were thus able to launch Electrum-NMC.
+
+## ncprop279 Support
+
+This one was easier.  ncprop279 doesn't need to write to any config directory, and it will happily look for a config file adjacent to its executable file if one is there.  So I just stuck `ncprop279.conf` in the same directory as the executable, and we were done there.
+
+## StemNS Support
+
+A minor tweak was needed for StemNS, so that it would search for `ncprop279` relative to StemNS's script directory instead of relative to the `PATH`.  Turns out that macOS doesn't standardize the `PATH` in the way that GNU/Linux and Windows do, so relying on the `PATH` wasn't usable on macOS.  Not too bad.
```

Comment:
Why is it not too bad that you can't rely on the `PATH` on macOS?

Upon re-reading, I realize the "not too bad" applies to how much work it was, but this should probably be re-worded so it's obvious the second sentence is related to the first